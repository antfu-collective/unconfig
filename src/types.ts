import type { Arrayable, Awaitable } from '@antfu/utils'

export const defaultExtensions = ['mts', 'cts', 'ts', 'mjs', 'cjs', 'js', 'json', '']

export type BuiltinParsers = 'require' | 'json'

export type CustomParser<T> = (filepath: string) => Awaitable<T |undefined>

export interface LoadConfigSource<T = any> {
  files: Arrayable<string>

  /**
   * @default ['mts', 'cts', 'ts', 'mjs', 'cjs', 'js', 'json', '']
   */
  extensions?: string[]

  /**
   * Loader for loading config,
   *
   * @default 'auto'
   */
  parser?: BuiltinParsers | CustomParser<T> | 'auto'

  /**
   * Rewrite the config object,
   * return nullish value to bypassing loading the file
   */
  rewrite?: <F = any>(obj: F, filepath: string) => Promise<T | undefined> | T | undefined

  /**
   * Transform the source code before loading,
   * return nullish value to skip transformation
   */
  transform?: (code: string, filepath: string) => Promise<string | undefined> | string | undefined

  /**
   * Skip this source if error occurred on loading
   *
   * @default false
   */
  skipOnError?: boolean
}

export interface SearchOptions {
  /**
   * Root directory
   *
   * @default process.cwd()
   */
  cwd?: string

  /**
   * @default path.parse(cwd).root
   */
  stopAt?: string

  /**
   * Load from multiple sources and merge them
   *
   * @default false
   */
  merge?: boolean
}

export interface LoadConfigOptions<T = any> extends SearchOptions {
  sources: Arrayable<LoadConfigSource<T>>
  defaults?: T
}

export interface LoadConfigResult<T> {
  config: T
  sources: Source[]
}

export interface Source {
  filepath: string
  source?: string
}
