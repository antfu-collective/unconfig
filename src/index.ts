import { promises as fs } from 'node:fs'
import { basename, dirname, join } from 'node:path'
import process from 'node:process'
import { notNullish, toArray } from '@antfu/utils'
import defu from 'defu'
import type { LoadConfigOptions, LoadConfigResult, LoadConfigSource } from './types'
import { defaultExtensions } from './types'
import { findUp } from './fs'
import { interopDefault } from './interop'

export * from './types'

export function createConfigLoader<T>(options: LoadConfigOptions) {
  const sources = toArray(options.sources || [])
  const {
    cwd = process.cwd(),
    merge,
    defaults,
  } = options

  const results: LoadConfigResult<T>[] = []
  let matchedFiles: [LoadConfigSource, string[]][] | undefined

  async function findConfigs() {
    if (matchedFiles == null)
      matchedFiles = []

    matchedFiles.length = 0
    for (const source of sources) {
      const { extensions = defaultExtensions } = source

      const flatTargets = toArray(source?.files || [])
        .flatMap(file => !extensions.length
          ? [file]
          : extensions.map(i => i ? `${file}.${i}` : file),
        )

      const files = await findUp(flatTargets, { cwd, stopAt: options.stopAt, multiple: merge })

      matchedFiles.push([source, files])
    }

    return matchedFiles.flatMap(i => i[1])
  }

  async function load(force = false): Promise<LoadConfigResult<T>> {
    if (matchedFiles == null || force)
      await findConfigs()

    for (const [source, files] of matchedFiles!) {
      if (!files.length)
        continue

      if (!merge) {
        const result = await loadConfigFile(files[0], source)
        if (result) {
          return {
            config: applyDefaults(result.config, defaults),
            sources: result.sources,
            dependencies: result.dependencies,
          }
        }
      }
      else {
        results.push(
          ...(await Promise.all(
            files.map(file => loadConfigFile(file, source)),
          )
          ).filter(notNullish),
        )
      }
    }

    if (!results.length) {
      return {
        config: defaults,
        sources: [],
      }
    }

    return {
      config: applyDefaults(...results.map(i => i.config), defaults),
      sources: results.map(i => i.sources).flat(),
      dependencies: results.flatMap(i => i.dependencies || []),
    }
  }

  return {
    load,
    findConfigs,
  }
}

function applyDefaults(...args: any[]): any {
  // defu does not support top-level array merging, we wrap it with an object and unwrap it
  // @ts-expect-error cast
  return defu(...args.map((i: any) => ({ config: i }))).config
}

export async function loadConfig<T>(options: LoadConfigOptions<T>): Promise<LoadConfigResult<T>> {
  return createConfigLoader<T>(options).load()
}

async function loadConfigFile<T>(filepath: string, source: LoadConfigSource<T>): Promise<LoadConfigResult<T> | undefined> {
  let config: T | undefined

  let parser = source.parser || 'auto'

  let bundleFilepath = filepath
  let code: string | undefined

  async function read() {
    if (code == null)
      code = await fs.readFile(filepath, 'utf-8')
    return code
  }

  if (source.transform) {
    const transformed = await source.transform(await read(), filepath)
    if (transformed) {
      bundleFilepath = join(dirname(filepath), `__unconfig_${basename(filepath)}`)
      await fs.writeFile(bundleFilepath, transformed, 'utf-8')
      code = transformed
    }
  }

  if (parser === 'auto') {
    try {
      config = JSON.parse(await read())
      parser = 'json'
    }
    catch {
      parser = 'import'
    }
  }

  let dependencies: string[] | undefined

  try {
    if (!config) {
      if (typeof parser === 'function') {
        config = await parser(filepath)
      }
      else if (parser === 'require' || parser === 'import') {
        config = await import('importx')
          .then(async (r) => {
            const mod = await r.import(bundleFilepath, {
              parentURL: filepath,
              cache: false,
            })
            dependencies = r.getModuleInfo(mod)?.dependencies
            return interopDefault(mod)
          })
      }
      else if (parser === 'json') {
        config = JSON.parse(await read())
      }
    }

    if (!config)
      return

    const rewritten = source.rewrite
      ? await source.rewrite(config, filepath)
      : config

    if (!rewritten)
      return undefined

    return {
      config: rewritten,
      sources: [filepath],
      dependencies,
    }
  }
  catch (e) {
    if (source.skipOnError)
      return
    throw e
  }
  finally {
    if (bundleFilepath !== filepath)
      await fs.unlink(bundleFilepath).catch()
  }
}
