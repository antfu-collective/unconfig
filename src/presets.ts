import type { Arrayable } from '@antfu/utils'
import { toArray } from '@antfu/utils'
import type { LoadConfigSource } from './types'

export interface SourceVitePluginConfigOptions {
  plugins: Arrayable<string>
  /**
   * Parameters that passed to when the default export is a function
   */
  parameters?: any[]
}

export interface SourceObjectFieldOptions extends Omit<LoadConfigSource, 'rewrite'> {
  fields: Arrayable<string>
  /**
   * Parameters that passed to when the default export is a function
   */
  parameters?: any[]
}

export interface SourcePluginFactoryOptions extends Omit<LoadConfigSource, 'transform'>{
  targetModule: string
  /**
   * Parameters that passed to when the default export is a function
   */
  parameters?: any[]
}

/**
 * Rewrite the config file and extract the options passed to plugin factory
 * (e.g. Vite and Rollup plugins)
 */
export function sourcePluginFactory(options: SourcePluginFactoryOptions) {
  return {
    ...options,
    transform: (source: string) => {
      const prefix = `
let __unconfig_data;
let __unconfig_stub = function (data) { __unconfig_data = data };
__unconfig_stub.default = (data) => { __unconfig_data = data };
`
      const suffix = 'export default __unconfig_data;'
      let code = source
        .replace(new RegExp(`import (.+?) from (['"])${options.targetModule}\\2`), 'const $1 = __unconfig_stub;')
        .replace('export default', 'const __unconfig_default = ')
      if (code.includes('__unconfig_default'))
        code += `\nif (typeof __unconfig_default === "function") __unconfig_default(...${JSON.stringify(options.parameters || [])});`
      return `${prefix}${code}${suffix}`
    },
  }
}

export function sourceVitePluginConfig(options: SourceVitePluginConfigOptions): LoadConfigSource {
  const plugins = toArray(options.plugins)
  return {
    files: ['vite.config'],
    async rewrite(obj) {
      const config = await (typeof obj === 'function' ? obj(...options.parameters || [{ env: {} }, {}]) : obj)
      if (!config)
        return config
      return config.plugins.find((i: any) => plugins.includes(i.name) && i?.api?.config)?.api?.config
    },
  }
}

/**
 * Get one field of the config object
 */
export function sourceObjectFields(options: SourceObjectFieldOptions): LoadConfigSource {
  const fields = toArray(options.fields)
  return {
    ...options,
    async rewrite(obj) {
      const config = await (typeof obj === 'function' ? obj(...options.parameters || []) : obj)
      if (!config)
        return config
      for (const field of fields) {
        if (field in config)
          return config[field]
      }
    },
  }
}

/**
 * Get one field of `package.json`
 */
export function sourcePackageJsonFields(options: Pick<SourceObjectFieldOptions, 'fields'>): LoadConfigSource {
  return sourceObjectFields({
    files: ['package.json'],
    extensions: [],
    parser: 'json',
    fields: options.fields,
  })
}
