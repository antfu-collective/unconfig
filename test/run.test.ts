import { resolve } from 'path'
import { expect, it } from 'vitest'
import { loadConfig } from '../src'
import { sourcePackageJsonFields, sourcePluginFactory } from '../src/presets'

it('loads the given sources', async() => {
  const result = await loadConfig({
    sources: [
      {
        files: 'un.config',
      },
      sourcePackageJsonFields({
        fields: 'un',
      }),
      sourcePluginFactory({
        targetModule: 'stub',
        files: 'rewrite.js',
        extensions: [],
      }),
    ],
    cwd: resolve(__dirname, 'fixtures'),
    defaults: {
      defaults: 'default',
      deep: { foo: 'hi' },
    },
    merge: true,
  })

  expect(result.config).toMatchSnapshot('config')
  expect(result.sources.length).toMatchSnapshot('files')
  expect(result.sources.map(({ source }) => source)).toMatchSnapshot('sources')
})
