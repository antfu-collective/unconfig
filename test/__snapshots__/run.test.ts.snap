// Vitest Snapshot v1

exports[`loads the given sources > config 1`] = `
{
  "deep": {
    "foo": "you should get this",
  },
  "defaults": "default",
  "from": "package.json",
  "importValue": "uwu",
  "json": "allo",
  "mjs": "hi",
  "rewrite": "Hi",
  "ts": "hello",
}
`;

exports[`loads the given sources > files 1`] = `5`;

exports[`loads the given sources > sources 1`] = `
[
  "\\"use strict\\";Object.defineProperty(exports, \\"__esModule\\", { value: true });exports.default = void 0;var _file = require(\\"./file\\");var _default =

{
  ts: 'hello',
  deep: {
    foo: 'you should get this' },

  importValue: _file.value };exports.default = _default;",
  "\\"use strict\\";Object.defineProperty(exports, \\"__esModule\\", { value: true });exports.default = void 0;var _default = {
  mjs: 'hi',
  deep: {
    foo: 'this should be override' } };exports.default = _default;",
  "{
  \\"json\\": \\"allo\\"
}
",
  "{
  \\"un\\": {
    \\"from\\": \\"package.json\\"
  }
}
",
  "\\"use strict\\";Object.defineProperty(exports, \\"__esModule\\", { value: true });exports.default = void 0;
let __unconfig_data;
let __unconfig_stub = function (data) {__unconfig_data = data;};
__unconfig_stub.default = (data) => {__unconfig_data = data;};
const a = __unconfig_stub;

const __unconfig_default = () => {
  return a({
    rewrite: 'Hi' });

};

if (typeof __unconfig_default === \\"function\\") __unconfig_default();var _default = __unconfig_data;exports.default = _default;",
]
`;
